doc/sphinx/proof-engine/ltac2.rst

```diff
@@ -467,23 +467,68 @@ destroy all backtrack and return values.
 Backtracking
 ++++++++++++
 
-In Ltac2, we have the following backtracking primitives, defined in the
-`Control` module::
+Ltac2 models backtracking computations using streams of values: the head of a non-empty stream represent the first path of computation while its tail stands for the alternative paths of computation when backtracking occurs. The empty stream holding no values represent failure and initiates backtracking.
+
+Backtracking failures (empty streams) are moreover decorated with exceptions to inform the reason of failure. These exceptions are then passed to the tail of a non-empty backtracking computation. The following Ltac2 type summarizes the model for backtracking computations
+
+  .. rocqtop:: in
+
+    Ltac2 Type rec 'a backtracking_stream :=
+      [ EmptyStream(exn)
+      | ConsStream('a, (exn -> 'a backtracking_stream)) ].
+
+where `EmptyStream` is a backtracking failure holding an exception and `ConsStream` is a backtracking success with a value `'a` as its head and a backtracking handler parameterized by an exception. When sequenced with additional backtracking computations, a backtracking success may lead to a failure with an exception that will be passed to the handler.
+
+
+In practice, the backtracking aspects of Ltac2's computations can be accessed through the following primitives, defined in the `Control` module::
 
   Ltac2 Type 'a result := [ Val ('a) | Err (exn) ].
 
   val zero : exn -> 'a
   val plus : (unit -> 'a) -> (exn -> 'a) -> 'a
   val case : (unit -> 'a) -> ('a * (exn -> 'a)) result
 
-If one views thunks as lazy lists, then `zero` is the empty list and `plus` is
-list concatenation, while `case` is pattern-matching.
+Informally, the primitive `zero e` ends the current computation and returns to the last backtracking point with exception `e`. In the stream model, it corresponds to the empty stream.
+The primitive `plus t handler` installs a backtracking point around `t`, using `handler` to handle any exception raised by a `zero` while evaluating `t`. This corresponds to appending two streams in the stream model.
```

Comment:
What does `plus t hanlder ; x` does if `x` fails ?

Reply:
cf the equation
~~~
  let x := plus t f in u ≡ plus (fun () => let x := t in u) (fun e => let x := f e in u)
~~~
below, considering `a; b` is equivalent to `let _ := a in b` for any expressions a and b